appinventor/blocklyeditor/src/backpack.js

```diff
@@ -325,21 +338,27 @@ Blockly.Backpack.prototype.addToBackpack = function(block, store) {
   block.setCollapsed(isCollapsed);
 
   // Add the block to the backpack
-  var bp_contents = this.getContents();
-  bp_contents.push("<xml>" + Blockly.Xml.domToText(xmlBlock) + "</xml>");
-  // We technically do not need to set the contents here since the contents are manipulated by
-  // reference, but separating the setting from modifying allows us to use different, non-in-memory
-  // storage in the future.
-  this.setContents(bp_contents, store);
-  this.grow();
-  Blockly.getMainWorkspace().playAudio('backpack');
-
-  // update the flyout when it's visible
-  if (this.flyout_.isVisible()) {
-    this.isAdded = true;
-    this.openBackpack();
-    this.isAdded = false;
-  }
+  var p = this;
+  this.getContents(function(bp_contents) {
```

Comment:
We should think about how this interacts with the Add All to Backpack menu item.

This will cause an asynchronous HTTP GET per top-level block. This might result in us reading multiple copies of the original backpack rather than read one, write one, .... In extreme circumstances the order of the backpack retrievals may be in the wrong order so we end up "undoing" a previous set. I would suggest adding a parameter to take an existing contents list to append entries. This way we can use the contents object retrieved in line 316 and then push it back once all of the entries have been added and the add all operation is atomic from the POV of the network.

Reply:
Actually, the AddAllToBackpack is just broken! I have fixed it and the next commit will fix it.